src/devices/src/virtio/block/device.rs

```diff
@@ -29,61 +29,93 @@ use super::{
+        let disk_size = disk_image.seek(SeekFrom::End(0))? as u64;
+
+        Ok(Self {
+            nsectors: disk_size / SECTOR_SIZE,
+            image_id: Self::build_disk_image_id(&disk_image),
+            file_path: disk_image_path,
+            file: disk_image,
+            size: disk_size,
+        })
+    }
+
+    fn build_device_id(disk_file: &File) -> result::Result<String, Error> {
+        let blk_metadata = disk_file.metadata().map_err(Error::GetFileMetadata)?;
+        // This is how kvmtool does it.
+        let device_id = format!(
+            "{}{}{}",
+            blk_metadata.st_dev(),
+            blk_metadata.st_rdev(),
+            blk_metadata.st_ino()
         );
+        Ok(device_id)
     }
-    let mut config = Vec::with_capacity(CONFIG_SPACE_SIZE);
-    let num_sectors = disk_size >> SECTOR_SHIFT;
-    for i in 0..CONFIG_SPACE_SIZE {
-        config.push((num_sectors >> (8 * i)) as u8);
+
+    fn build_disk_image_id(disk_file: &File) -> Vec<u8> {
+        let mut default_id = vec![0; VIRTIO_BLK_ID_BYTES as usize];
+        match Self::build_device_id(disk_file) {
+            Err(_) => {
+                warn!("Could not generate device id. We'll use a default.");
+            }
+            Ok(m) => {
+                // The kernel only knows to read a maximum of VIRTIO_BLK_ID_BYTES.
+                // This will also zero out any leftover bytes.
+                let disk_id = m.as_bytes();
+                let bytes_to_copy = cmp::min(disk_id.len(), VIRTIO_BLK_ID_BYTES as usize);
+                default_id[..bytes_to_copy].clone_from_slice(&disk_id[..bytes_to_copy])
+            }
+        }
+        default_id
     }
-    config
-}
 
-fn build_device_id(disk_image: &File) -> result::Result<String, Error> {
-    let blk_metadata = disk_image.metadata().map_err(Error::GetFileMetadata)?;
-    // This is how kvmtool does it.
-    let device_id = format!(
-        "{}{}{}",
-        blk_metadata.st_dev(),
-        blk_metadata.st_rdev(),
-        blk_metadata.st_ino()
-    );
-    Ok(device_id)
-}
+    /// Backing file path.
+    pub fn file_path(&self) -> &String {
+        &self.file_path
+    }
 
-fn build_disk_image_id(disk_image: &File) -> Vec<u8> {
-    let mut default_disk_image_id = vec![0; VIRTIO_BLK_ID_BYTES as usize];
-    match build_device_id(disk_image) {
-        Err(_) => {
-            warn!("Could not generate device id. We'll use a default.");
+    /// Provides vec containing the virtio block configuration space
+    /// buffer. The config space is populated with the disk size based
+    /// on the backing file size.
+    pub fn virtio_block_config_space(&self) -> Vec<u8> {
+        // We only support disk size, which uses the first two words of the configuration space.
+        // If the image is not a multiple of the sector size, the tail bits are not exposed.
+        // The config space is little endian.
+        if self.size % SECTOR_SIZE != 0 {
```

Comment:
I would move this check to the constructor and then we can just keep `nsectors` and drop the `size` field.

Reply:
Right, done.